fe/src/main/java/org/apache/doris/common/util/DynamicPartitionUtil.java

```diff
@@ -273,34 +321,110 @@ public static String getPartitionFormat(Column column) throws DdlException {
         if (timeUnit.equalsIgnoreCase(TimeUnit.DAY.toString())) {
-            calendar.add(Calendar.DAY_OF_MONTH, offset);
+            return getPartitionRangeOfDay(current, offset, tz, format);
         } else if (timeUnit.equalsIgnoreCase(TimeUnit.WEEK.toString())) {
-            calendar.add(Calendar.WEEK_OF_MONTH, offset);
-        } else {
-            calendar.add(Calendar.MONTH, offset);
+            return getPartitionRangeOfWeek(current, offset, property.getStartOfWeek(), tz, format);
+        } else { // MONTH
+            return getPartitionRangeOfMonth(current, offset, property.getStartOfMonth(), tz, format);
+        }
+    }
+    
+    /*
+     * return formatted string of partition range in DAY granularity.
+     * offset: The offset from the current day. 0 means current day, 1 means tomorrow, -1 means yesterday.
+     * format: the format of the return date string.
+     * 
+     * Eg:
+     *  Today is 2020-05-24, offset = -1
+     *  It will return 2020-05-23
+     */
+    private static String getPartitionRangeOfDay(Calendar current, int offset, TimeZone tz, String format) {
+        current.add(Calendar.DATE, offset);
+        return getFormattedTimeWithoutHourMinuteSecond(current, format);
+    }
+
+    /*
+     * return formatted string of partition range in WEEK granularity.
+     * offset: The offset from the current week. 0 means current week, 1 means next week, -1 means last week.
+     * startOf: Define the start day of each week. 1 means MONDAY, 7 means SUNDAY.
+     * format: the format of the return date string.
+     * 
+     * Eg:
+     *  Today is 2020-05-24, offset = -1, startOf.dayOfWeek = 3
+     *  It will return 2020-05-20  (Wednesday of last week)
+     */
+    private static String getPartitionRangeOfWeek(Calendar current, int offset, StartOfDate startOf, TimeZone tz,
+            String format) {
+        Preconditions.checkArgument(startOf.isStartOfWeek());
+        // 1. get the offset week
+        current.add(Calendar.WEEK_OF_YEAR, offset);
+        // 2. get the date of `startOf` week
+        int day = current.get(Calendar.DAY_OF_WEEK);
+        // SUNDAY will return 1, we will set it to 7, and make MONDAY to 1, and so on
+        day = day == 1 ? 7 : day - 1;
+        current.add(Calendar.DATE, (startOf.dayOfWeek - day));
+        return getFormattedTimeWithoutHourMinuteSecond(current, format);
+    }
+
+    /*
+     * return formatted string of partition range in MONTH granularity.
+     * offset: The offset from the current month. 0 means current month, 1 means next month, -1 means last month.
+     * startOf: Define the start date of each month. 1 means start on the 1st of every month.
+     * format: the format of the return date string.
+     * 
+     * Eg:
+     *  Today is 2020-05-24, offset = 1, startOf.month = 3
+     *  It will return 2020-06-03 
+     */
+    private static String getPartitionRangeOfMonth(Calendar current, int offset, StartOfDate startOf, TimeZone tz,
```

Comment:
why tz is no use?

Reply:
What about use `LocalDateTime/ZonedDateTime` to replace all `Calendar`? They are thread-safe and support timezone

Reply:
Maybe in next PR? This PR is already too big

Reply:
OK, we can change it in next PR~